pyo3-macros-backend/src/pymethod.rs

```diff
@@ -572,12 +572,11 @@ pub fn impl_py_setter_def(
             _slf: *mut _pyo3::ffi::PyObject,
             _value: *mut _pyo3::ffi::PyObject,
         ) -> _pyo3::PyResult<::std::os::raw::c_int> {
-            let _value = py
-                .from_borrowed_ptr_or_opt(_value)
+            let _value = _pyo3::Bound::from_borrowed_ptr_or_opt(py, _value)
```

Comment:
I think we can avoid a reference-counting operation here by going via the `_pyo3::impl_::pymethods::BoundRef` machinery which is already used in a couple other places in these macros.

But instead of `BoundRef::ref_from_ptr` I think this will require introduction of `ref_from_ptr_or_opt`. I was about to do that as part of #3708, maybe I'll split it out quickly.

Then extract_bound below you can just use `::std::convert::Into::into` to convert into normal `&Bound`.

Reply:
Actually I didn't need it in #3708, but perhaps this code here is useful (to add to `Bound` implementation):

```rust
    /// Variant of the above which returns `None` for null pointers.
    ///
    /// # Safety
    /// - `ptr` must be a valid pointer to a Python object for the lifetime `'a, or null.
    #[inline]
    pub(crate) unsafe fn ref_from_ptr_or_opt<'a>(
        _py: Python<'py>,
        ptr: &'a *mut ffi::PyObject,
    ) -> &'a Option<Self> {
        &*(ptr as *const *mut ffi::PyObject).cast::<Option<Bound<'py, PyAny>>>()
    }
```

Reply:
I'm happy to try this, or you can include it when adding `ref_from_ptr_or_opt`.

Reply:
I think I won't be getting round to adding `ref_from_ptr_or_opt`, so perhaps include it in this PR for now?